File: Data-Structure/All-In-One-LeetCode/00056_Merge_Intervals/merge.js
const merge = arr => {
    if (arr.length === 1) return [arr[0]];
    let N = arr.length;
    arr.sort((a, b) => a[0] - b[0]);
    let res = [arr[0]];

    for (let i = 1; i < N; i++) {
        let last = res[res.length - 1];
        let curr = arr[i];

        if (last[1] >= curr[0]) {
            last[1] = Math.max(last[1], curr[1]);
        } else {
            res.push(curr);
        }
    }
    return res;
}

File: Data-Structure/_LeetCode_After_500_Done/Top Facebook Questions/56. Merge Intervals/Merge.java
import java.util.*;
import java.lang.*;

public class Merge {
    private List<int[]> res = new ArrayList<int[]>();

    public int[][] merge(int[][] arr) {
        int N = arr.length;
        if (N < 2)
            return arr;
        Arrays.sort(arr, (a, b) -> a[0] - b[0]);
        res.add(arr[0]);
        for (int i = 1; i < N; i++) {
            if (res.get(res.size() - 1)[1] >= arr[i][0]) {
                res.get(res.size() - 1)[1] = Math.max(res.get(res.size() - 1)[1], arr[i][1]);
            } else {
                res.add(arr[i]);
            }
        }
        return res.toArray(new int[res.size()][]);
    }

    public static void main(String[] args) {
        int[][] intervals = new int[][] { { 1, 3 }, { 2, 6 }, { 8, 10 }, { 15, 18 } };
        Merge test = new Merge();
        int[][] res = test.merge(intervals);

        for (int i = 0; i < res.length; i++) {
            System.out.println(res[i][0] + "," + res[i][1]);
        }

    }
}
